pkg/controller/endpoint/endpoints_controller_test.go

```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
 	}
 }
 
+func TestMultipleServiceChanges(t *testing.T) {
+	mux := http.NewServeMux()
+	ns := metav1.NamespaceDefault
+	expectedSubsets := []v1.EndpointSubset{{
+		Addresses: []v1.EndpointAddress{
+			{IP: "1.2.3.4", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod0", Namespace: ns}},
+		},
+	}}
+	endpoint := &v1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: ns, ResourceVersion: "1"},
+		Subsets:    expectedSubsets,
+	}
+
+	var endpointsStore cache.Store
+	var endpoints *endpointController
+	blockDelete := make(chan struct{})
+	blockNextAction := make(chan struct{})
+	stopChan := make(chan struct{})
+
+	handlerFunc := func(res http.ResponseWriter, req *http.Request) {
+		if endpointsStore == nil {
+			res.WriteHeader(http.StatusInternalServerError)
+			res.Write([]byte("endpointsStore has not been set yet"))
+			return
+		}
+		if endpoints == nil {
+
+			res.WriteHeader(http.StatusInternalServerError)
+			res.Write([]byte("endpoints"))
+			return
+		}
+
+		if req.Method == "POST" {
+			endpointsStore.Add(endpoint)
+			blockNextAction <- struct{}{}
+		}
+
+		if req.Method == "DELETE" {
+			go func() {
+				// Delay the deletion of endoints to make endpoint cache out of sync
+				<-blockDelete
+				endpointsStore.Delete(endpoint)
+				endpoints.onEndpointsDelete(endpoint)
+			}()
+			blockNextAction <- struct{}{}
+		}
+
+		res.WriteHeader(http.StatusOK)
+		res.Write([]byte(runtime.EncodeOrDie(clientscheme.Codecs.LegacyCodec(v1.SchemeGroupVersion), &v1.Endpoints{})))
+	}
+
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/endpoints", handlerFunc)
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/endpoints/", handlerFunc)
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/events", func(res http.ResponseWriter, req *http.Request) {})
+	mux.HandleFunc("/", func(res http.ResponseWriter, req *http.Request) {
+		t.Errorf("unexpected request: %v", req.RequestURI)
+		http.Error(res, "", http.StatusNotFound)
+	})
+	testServer := httptest.NewServer(mux)
+
+	defer testServer.Close()
+	endpoints = newController(testServer.URL, 0*time.Second)
+	endpointsStore = endpoints.endpointsStore
+
+	addPods(endpoints.podStore, ns, 1, 1, 0, ipv4only)
+	go func() { endpoints.Run(1, stopChan) }()
+
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: ns},
+		Spec: v1.ServiceSpec{
+			Selector:  map[string]string{"foo": "bar"},
+			ClusterIP: "None",
+			Ports:     nil,
+		},
+	}
+
+	endpoints.serviceStore.Add(svc)
+	endpoints.onServiceUpdate(svc)
+	timer := time.NewTimer(1 * time.Second)
```

Comment:
Can we always guarantee that 1 second will be enough here?

Reply:
Looking at the other tests, when waiting for the queue to be processed 1 second seemed to be used. But if there is a chance of increased flakiness, I can increase the timeout. What would you suggest?

Reply:
👍 That does seem reasonable. The other tests have not been flaky.